subsys/net/lib/azure_fota/azure_fota.c

```diff
@@ -0,0 +1,546 @@
+		goto clean_exit;
+	}
+
+	if (cJSON_IsString(fw_version_obj)) {
+		if (strlen(fw_version_obj->valuestring) >=
+		    sizeof(fota_object.version)) {
+			LOG_ERR("Version string longer than buffer");
+			err = -ENOMEM;
+			goto clean_exit;
+		}
+
+		memcpy(fota_object.version, fw_version_obj->valuestring,
+		       sizeof(fota_object.version));
+
+		LOG_DBG("Incoming firmware version: %s",
+			log_strdup(fota_object.version));
+	} else if (cJSON_IsNumber(fw_version_obj)) {
+		ssize_t len = snprintk(fota_object.version,
+				       sizeof(fota_object.version),
+				       "%d", fw_version_obj->valueint);
+
+		if ((len < 0) || (len >= sizeof(fota_object.version))) {
+			LOG_ERR("Version number too long for buffer");
+			err = -ENOMEM;
+		}
+	} else {
+		LOG_DBG("Invalid fwVersion format received");
+		goto clean_exit;
+	}
+
+	/* Firmware image location */
+	fw_location_obj = cJSON_GetObjectItemCaseSensitive(fw_obj,
+							   "fwLocation");
+	if (fw_location_obj == NULL) {
+		LOG_DBG("No fwLocation object found");
+		goto clean_exit;
+	}
+
+	/* Firmware image hostname */
+	fw_hostname_obj = cJSON_GetObjectItemCaseSensitive(fw_location_obj,
+							   "host");
+	if (fw_hostname_obj == NULL) {
+		LOG_DBG("No hostname object found");
+		goto clean_exit;
+	}
+
+	if (strlen(fw_hostname_obj->valuestring) >=
+	    sizeof(fota_object.host)) {
+		LOG_ERR("Received hostname larger (%d) than buffer (%d)",
+			strlen(fw_hostname_obj->valuestring),
+			sizeof(fota_object.host));
+		goto clean_exit;
+	}
+
+	strncpy(fota_object.host, fw_hostname_obj->valuestring,
+		sizeof(fota_object.host));
+
+	/* Firmware image file path */
+	fw_path_obj = cJSON_GetObjectItemCaseSensitive(fw_location_obj, "path");
+	if (fw_path_obj == NULL) {
+		LOG_DBG("No path object found");
+		goto clean_exit;
+	}
+
+	if (strlen(fw_path_obj->valuestring) >= sizeof(fota_object.path)) {
+		LOG_ERR("Received file path larger (%d) than buffer (%d)",
+			strlen(fw_path_obj->valuestring),
+			sizeof(fota_object.path));
+		goto clean_exit;
+	}
+
+	strncpy(fota_object.path, fw_path_obj->valuestring,
+		sizeof(fota_object.path));
+
+	/* Firmware fragment size */
+	fw_fragment_obj = cJSON_GetObjectItemCaseSensitive(fw_obj,
+							"fwFragmentSize");
+	if (fw_fragment_obj == NULL) {
+		LOG_DBG("Failed to parse 'fwFragmentSize' object");
+		goto clean_exit;
+	}
+
+	fota_object.fragment_size = fw_fragment_obj->valueint;
+
+	err = 0;
+
+clean_exit:
+	cJSON_Delete(root_obj);
+
+	return err;
+}
+
+/* Public functions */
+int azure_fota_init(azure_fota_callback_t evt_handler)
+{
+	int err;
+
+	if (evt_handler == NULL) {
+		return -EINVAL;
+	}
+
+	callback = evt_handler;
+
+	err = fota_download_init(fota_dl_handler);
+	if (err != 0) {
+		LOG_ERR("fota_download_init error %d", err);
+		return err;
+	}
+
+	cJSON_Init();
+
+	/* Get current firmware information, report previous update? */
+
+	fota_state = STATUS_CURRENT;
+
+	LOG_DBG("Current version: %s", log_strdup(get_current_version()));
+
+	return 0;
```

Comment:
Could return err here.

Reply:
I think returning `0` makes it a bit clearer that twe would never reach here if an error occured, and an extra bonus is a tighter scope of `err` and save one load from memory